crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -0,0 +1,787 @@
+use crate::{
+    database::{
+        database_description::{
+            DatabaseDescription,
+            DatabaseHeight,
+        },
+        Error as DatabaseError,
+        Result as DatabaseResult,
+    },
+    state::{
+        historical_rocksdb::{
+            description::Historical,
+            modifications_history::ModificationsHistory,
+            view_at_height::ViewAtHeight,
+        },
+        iterable_key_value_view::IterableKeyValueViewWrapper,
+        key_value_view::KeyValueViewWrapper,
+        rocks_db::RocksDb,
+        ColumnType,
+        IterableKeyValueView,
+        KeyValueView,
+        TransactableStorage,
+    },
+};
+use fuel_core_storage::{
+    iter::{
+        BoxedIter,
+        IterDirection,
+        IterableStore,
+        IteratorOverTable,
+    },
+    kv_store::{
+        KVItem,
+        KeyValueInspect,
+        Value,
+        WriteOperation,
+    },
+    not_found,
+    transactional::{
+        Changes,
+        ConflictPolicy,
+        ReadTransaction,
+        StorageTransaction,
+    },
+    Error as StorageError,
+    Result as StorageResult,
+    StorageAsMut,
+};
+use itertools::Itertools;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+use std::num::NonZeroU64;
+
+pub mod description;
+pub mod modifications_history;
+pub mod view_at_height;
+
+#[derive(Copy, Clone, Default, Debug, Eq, PartialEq)]
+/// Defined policies for the state rewind behaviour of the database.
+pub enum StateRewindPolicy {
+    #[default]
+    /// The checkpoint will be created only for the latest height.
+    NoRewind,
+    /// The checkpoint will be created for each height.
+    RewindFullRange,
+    /// The checkpoint will be created for each height
+    /// in the range `[latest_height-size..latest_height]`.
+    RewindRange { size: NonZeroU64 },
+}
+
+#[derive(Debug)]
+pub struct HistoricalRocksDB<Description> {
+    state_rewind_policy: StateRewindPolicy,
+    db: RocksDb<Description>,
+    history: RocksDb<Historical<Description>>,
+}
+
+impl<Description> HistoricalRocksDB<Description>
+where
+    Description: DatabaseDescription,
+{
+    pub fn new(
+        db: RocksDb<Description>,
+        state_rewind_policy: StateRewindPolicy,
+    ) -> DatabaseResult<Self> {
+        let path = db.path().join("history");
+        let history = RocksDb::default_open(path, None)?;
+        Ok(Self {
+            state_rewind_policy,
+            db,
+            history,
+        })
+    }
+
+    fn reverse_history_changes(&self, changes: &Changes) -> StorageResult<Changes> {
+        let mut reverse_changes = Changes::default();
+
+        for (column, column_changes) in changes {
+            let results = self
+                .db
+                .multi_get(*column, column_changes.iter().map(|(k, _)| k))?;
```

Comment:
Could there be potential race conditions here, ie such as many changes being committed to in parallel during synchronization?

Reply:
If you work with pure `HistoricalRocksDB` - yes, multiple commits can produce different results.

If you work with `GenericDatabase` - it guarantees only one commit per time.